src/jit/lowerxarch.cpp

```diff
@@ -135,190 +135,144 @@ void Lowering::LowerStoreIndir(GenTreeIndir* node)
-//
-// Return Value:
-//    None.
+//    blkNode - The block store node to lower
 //
 void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
 {
-    GenTree* dstAddr       = blkNode->Addr();
-    unsigned size          = blkNode->Size();
-    GenTree* source        = blkNode->Data();
-    GenTree* srcAddrOrFill = nullptr;
-    bool     isInitBlk     = blkNode->OperIsInitBlkOp();
+    GenTree* dstAddr = blkNode->Addr();
+    GenTree* src     = blkNode->Data();
+    unsigned size    = blkNode->Size();
 
-    if (!isInitBlk)
+    if (blkNode->OperIsInitBlkOp())
     {
-        // CopyObj or CopyBlk
-        if (blkNode->OperIs(GT_STORE_OBJ) && (!blkNode->AsObj()->GetLayout()->HasGCPtr() || blkNode->gtBlkOpGcUnsafe))
+        if (src->OperIs(GT_INIT_VAL))
         {
-            blkNode->SetOper(GT_STORE_BLK);
+            src->SetContained();
+            src = src->AsUnOp()->gtGetOp1();
         }
-        if (source->gtOper == GT_IND)
-        {
-            srcAddrOrFill = blkNode->Data()->gtGetOp1();
-        }
-    }
 
-    if (isInitBlk)
-    {
-        GenTree* initVal = source;
-        if (initVal->OperIsInitVal())
+        if (blkNode->OperIs(GT_STORE_OBJ))
         {
-            initVal->SetContained();
-            initVal = initVal->gtGetOp1();
+            blkNode->SetOper(GT_STORE_BLK);
         }
-        srcAddrOrFill = initVal;
-        // If we have an InitBlk with constant block size we can optimize several ways:
-        // a) If the size is smaller than a small memory page but larger than INITBLK_UNROLL_LIMIT bytes
-        //    we use rep stosb since this reduces the register pressure in LSRA and we have
-        //    roughly the same performance as calling the helper.
-        // b) If the size is <= INITBLK_UNROLL_LIMIT bytes and the fill byte is a constant,
-        //    we can speed this up by unrolling the loop using SSE2 stores.  The reason for
-        //    this threshold is because our last investigation (Fall 2013), more than 95% of initblks
-        //    in our framework assemblies are actually <= INITBLK_UNROLL_LIMIT bytes size, so this is the
-        //    preferred code sequence for the vast majority of cases.
-
-        // This threshold will decide from using the helper or let the JIT decide to inline
-        // a code sequence of its choice.
-        unsigned helperThreshold = max(INITBLK_STOS_LIMIT, INITBLK_UNROLL_LIMIT);
 
-        // TODO-X86-CQ: Investigate whether a helper call would be beneficial on x86
-        if (!blkNode->OperIs(GT_STORE_DYN_BLK) && size <= helperThreshold)
+        if (!blkNode->OperIs(GT_STORE_DYN_BLK) && (size <= INITBLK_UNROLL_LIMIT))
         {
-            // Always favor unrolling vs rep stos.
-            if (size <= INITBLK_UNROLL_LIMIT && initVal->IsCnsIntOrI())
+            if (!src->OperIs(GT_CNS_INT))
             {
+                // TODO-CQ: We could unroll even when the initialization value is not a constant
```

Comment:
While this is in lowerxarch it is not arch specific, you can always insert a `mul x, 0x01010101` to deal with non-const initialization. Not sure if anyone is doing this, though since `initblk` was exposed in the `Unsafe` class perhaps someone will complain about it.